changelog/ob.md

```diff
@@ -0,0 +1,282 @@
+dereferenced pointers are pointing to a valid memory object.
+
+
+## Scope of Prototype OB System
+
+This is a prototype OB system adapted to D. It is initially for pointers
+only, not dynamic arrays, class references, refs, or pointer fields
+of aggregates. Adding support for such adds complexity,
+but does not change the nature of it, hence it is deferred to later.
+RAII objects can safely manage their own memory, so are not covered
+by OB. Whether a pointer is allocate memory using the GC or some other
+storage allocator is immaterial to OB, they are not distinguished and are
+handled identically.
+
+The system is only active in functions annotated with the `@live` attribute.
+It is applied after semantic processing is done as purely a check
+for violations of the OB rules. No new syntax is added. No change is made
+to the code generated.
+If `@live` functions
+call non-`@live` functions, those called functions are expected to present
+an `@live` compatible interface, although it is not checked.
+
+The OB system will detect as errors:
+
+* dereferencing pointers that are in an invalid state
+* more than one active pointer to a mutable memory object
+
+It will not detect attempts to dereference `null` pointers or possibly
+`null` pointers. This is unworkable because there is no current method
+of annotating a type as a non-`null` pointer.
+
+
+## Core OB Principle
+
+The OB design follows from the following principle:
+
+For each memory object, there can exist either exactly one mutating pointer
+to it, or multiple non-mutating (read-only) pointers.
+
+
+## Design
+
+The single mutating pointer is called the "owner" of the memory object.
+It transitively owns the memory object and all memory objects accessible
+from it (i.e. the memory object graph). Since it is the sole pointer to
+that memory object, it can safely
+manage the memory (change its shape, allocate, free and resize) without
+pulling the rug out from under any other pointers (mutating or not)
+that may point to it.
+
+If there are multiple read-only pointers to the memory object graph,
+they can safely read from it without being concerned about the memory
+object graph being changed underfoot.
+
+The rest of the design is concerned with how pointers become owners,
+read only pointers, and invalid pointers, and how the Core OB Principle
+is maintained at all times.
+
+
+### Tracked Pointers
+
+The only pointers that are tracked are those declared in the `@live` function
+as `this`, function parameters or local variables. Variables from other
+functions are not tracked, even `@live` ones, as the analysis of interactions
+with other functions depends
+entirely on that function signature, not its internals.
+Parameters that are const are not tracked.
+
+
+### Pointer States
+
+Each pointer is in one of the following states:
+
+1. Undefined
+
+The pointer is in an invalid state. Dereferencing such a pointer is
+an error.
+
+2. Owner
```

Comment:
What happens if an owner pointer is assigned to another owner pointer?

```
int* p = cast(int*) malloc(int.sizeof);
int* q = p;  // ???

```

Reply:
then `p` enters an undefined state.

Reply:
I think this needs to be clarified in the DIP.